proposals/0166-dynamic-stack-frames.md

```diff
@@ -0,0 +1,154 @@
+a feature we expect to enable soon. 
+
+The changes proposed in this document would allow us to optimize stack memory 
+usage and remove the fragile stack frame gaps system. Note that we do not 
+propose to remove the existing maximum stack space limit: stack space stays 
+unchanged, what changes is how it is partitioned internally.
+
+## Alternatives Considered
+
+To cope with the SBF limitation of 4096 bytes for the frame size, we could 
+have increased such a number. Even though this would solve the original 
+problem, it would supply an unnecessary amount of memory to functions even 
+when they do not need them. In addition, such a solution would increase 
+pressure on the total memory available for the call stack. Either we would 
+need to increase the total allocation for the virtual machine or decrease the 
+maximum call depth.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+Bringing dynamic stack frames to the Solana Bytecode Format and its 
+corresponding virtual machine entails changes in several aspects of the 
+execution environment.
+
+### SBF architecture modifications
+
+
+We will introduce a new register R11 in the virtual machine, which is going 
+to hold the stack pointer. The program can only write to such a register and 
+modify it through the `add64 reg, imm` instruction. The verifier will enforce 
+these constraints on deployed programs. For further information about the 
+changes in the ISA, refer to the [SPF spec document](https://github.com/solana-labs/rbpf/blob/main/doc/bytecode.md).
+
+The R11 register will work in tandem with the R10 (frame pointer) register. 
+The former is write-only to the program, and the latter is read-only to the 
+program, forming a common design pattern in hardware engineering. More 
+details of this usage are in the following section.
+
+### Changes in the execution environment
+
+The R10 register will continue to hold the frame pointer, but we will manage 
+it differently. With fixed frames, when there is a function call we add 4096 
+to R10 and subtract it when the function returns. In the new scheme, we will 
+assign the value of R11 to R10 at function calls, and save R10’s former value 
+so that we can restore it when the function returns.
+
+The introduction of dynamic stack frames will change the direction of stack 
+growth. Presently, we stack frames on top of each other, but the memory usage 
+in them grows downward. In the new frame setting, both the placement of new 
+frames and the memory usage inside frames will be downward.
+
+The stack frame gaps feature, which creates a memory layout where frames are 
+interleaved with equally sized gaps, are not compatible with dynamic stack 
+frames and will be deactivated.
+
+### Changes in code generation
+
+In the compiler side, dynamic stack frames allow for some optimizations. 
+First, when a function does not need any stack allocated variable, code 
+generation will not create any instruction to modify R11. In addition, we 
```

Comment:
Won't we need to clear R11 before the call here?

Reply:
I guess you'd clear it as part of the call procedure so the callee won't have to deal with it

Reply:
R11 maintains a valid pointer to the stack throughout the program execution. There is no need to clean anything. If functions need stack space, they first decrement R11 with `add64 R11, -N`, and increment it when returning with `add64 R11, N`, where `N` is the number of stack space a function needs.